File: app/src/main/java/com/github/adrijanrogan/etiketa/ui/browse/BrowseViewModel.java
package com.github.adrijanrogan.etiketa.ui.browse;

import android.content.Context;
import android.os.Environment;

import com.github.adrijanrogan.etiketa.R;
import com.github.adrijanrogan.etiketa.util.FileComparator;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import androidx.lifecycle.ViewModel;

public class BrowseViewModel extends ViewModel {

    // Vedno hranimo referenco parent, da se lahko enostavneje vrnemo v visjo hierarhijo.
    // Ce sta parent in root enaka, smo ze najvisje.
    // Uporabnik se v visjo hierarhijo vraca s tipko nazaj, v primeru, da smo ze najvisje,
    // pa se aplikacija (kot je obicajno za tipko nazaj) zapre.
    private File root;
    private File parent;
    private File[] children;

    public BrowseViewModel() {
        super();
        root = new File(Environment.getExternalStorageDirectory().getPath() + "/");
        parent = root;
        children = parent.listFiles();
    }

    void goUp() {
        parent = parent.getParentFile();
        children = parent.listFiles();
    }

    void goDown(File file) {
        parent = file;
        children = parent.listFiles();
    }

    File[] getChildren() {
        return children;
    }

    boolean checkIfParentRoot() {
        return parent.getAbsolutePath().equals(root.getAbsolutePath());
    }

    String getTitle(Context context) {
        if (checkIfParentRoot()) {
            return context.getString(R.string.internal_storage);
        } else {
            return parent.getName();
        }
    }

    void sortFiles() {
        Arrays.sort(children, new FileComparator());
    }

    void removeHiddenFiles() {
        List<File> fileList = new ArrayList<>(Arrays.asList(children));
        for (int i = fileList.size() - 1; i >= 0; i--) {
            File file = fileList.get(i);
            if (file.getName().startsWith(".")) {
                fileList.remove(i);
            }
        }
        children = fileList.toArray(new File[0]);
    }
}

File: app/src/main/java/com/github/adrijanrogan/etiketa/ui/browse/BrowseViewModel.kt
package com.github.adrijanrogan.etiketa.ui.browse

import androidx.lifecycle.LiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import com.github.adrijanrogan.etiketa.repository.FileRepository
import com.github.adrijanrogan.etiketa.util.FileComparator
import java.io.File

class BrowseViewModel : ViewModel() {

    private val repository = FileRepository()

    fun getTree(): LiveData<List<File>> {
        return repository.getTree()
    }

    fun getFiles(showHidden: Boolean = false, sortOrder: Int = FileComparator.SORT_MODE_FILENAME):
            LiveData<List<File>> {
        val filesLiveData = repository.getFiles()
        return Transformations.map(filesLiveData) { files ->
            transformFiles(files, showHidden, sortOrder) }
    }

    private fun transformFiles(files: List<File>, showHidden: Boolean, sortOrder: Int): List<File> {
        val newList: MutableList<File> = ArrayList(files.size)
        if (!showHidden) {
            for (file in files) {
                if (!file.isHidden) newList.add(file)
            }
        } else {
            newList.addAll(files)
        }

        newList.sortWith(FileComparator(sortOrder))
        return newList
    }

    fun isRoot(): Boolean {
        return repository.checkIfRoot()
    }

    fun toParentFiles() {
        repository.toParentFiles()
    }

    fun toChildrenFiles(file: File) {
        repository.toChildrenFiles(file)
    }
}